src/render/src/lib.rs

```diff
@@ -146,70 +140,259 @@ pub mod shade;
 pub mod macros;
 */
 
+use std::collections::VecDeque;
+use core::{CommandQueue, QueueType, Surface, Swapchain, Device as CoreDevice};
+use core::pool::{CommandPool, CommandPoolCreateFlags};
+use core::format::RenderFormat;
+use memory::Typed;
+
+struct Queue<B: Backend, C> {
+    inner: CommandQueue<B, C>,
+    command_pool_receiver: encoder::CommandPoolReceiver<B, C>,
+    command_pool_sender: encoder::CommandPoolSender<B, C>,
+}
+
+impl<B: Backend, C> Queue<B, C> {
+    fn new(inner: CommandQueue<B, C>) -> Self {
+        let (command_pool_sender, command_pool_receiver) =
+            encoder::command_pool_channel();
+        Queue { inner, command_pool_sender, command_pool_receiver }
+    }
+
+    fn acquire_encoder_scope(&mut self) -> encoder::Scope<B, C> {
+        let initial_capacity = 4;
+        let flags = CommandPoolCreateFlags::empty();
+        let pool = self.command_pool_receiver.try_recv()
+            .map(|mut recycled| {
+                recycled.reset();
+                recycled
+            })
+            .unwrap_or_else(|_| {
+                CommandPool::from_queue(&self.inner, initial_capacity, flags)
+            });
+        encoder::Scope::new(pool, self.command_pool_sender.clone())
+    }
+}
+
+/* TODO
 pub(crate) enum Queue<B: Backend> {
     General(CommandQueue<B, core::General>),
     Graphics(CommandQueue<B, core::Graphics>),
 }
+*/
 
-pub struct Gpu<B: Backend> {
+// TODO: could be named Instance too ?
+pub struct Context<B: Backend> {
+    surface: B::Surface,
     device: Device<B>,
-    queue: Queue<B>,
+    queue: Queue<B, core::Graphics>,
+    swapchain: B::Swapchain,
+    frame_bundles: VecDeque<FrameBundle<B>>,
+    frame_acquired: Option<FrameBundle<B>>,
     garbage: handle::GarbageReceiver<B>,
 }
 
-impl<B: Backend> Gpu<B> {
-    pub fn new<A: core::Adapter<B>>(adapter: &A) -> Self {
-        // TODO: filter for queues which can be used for presentation with the display surface.
-        // TODO: use shorter core helpers
-        let queue_descs = adapter.get_queue_families()
-            .iter()
-            .map(|&(ref family, qtype)| (family, qtype, family.num_queues()) )
-            .collect::<Vec<_>>();
-        
+pub struct Backbuffer<B: Backend, Cf: RenderFormat> {
+    color: handle::RenderTargetView<B, Cf>,
+    // TODO: depth
+}
+
+use self::SyncState::*;
+#[derive(PartialEq)]
+enum SyncState {
```

Comment:
What's the deal with these sync states?

Reply:
The names might need some work but I use it to track synchronisation state at runtime.
For semaphores:
`Unreached` means that it has not been setup.
`Reached` means that it has been setup.
For fences:
`Unreached` means that the fence has not been waited upon.
`Reached` means that the fence has been waited upon.

It's also probably not necessary for all three primitives used here but it's easy to remove where not needed.